core/reactor_common.c

```diff
@@ -1571,20 +1571,39 @@ bool _lf_check_deadline(self_base_t* self, bool invoke_deadline_handler) {
  */
 void _lf_invoke_reaction(reaction_t* reaction, int worker) {
     
-    #ifdef NUMBER_OF_WORKERS
+    #ifdef LF_MULTI_THREADED
     lf_sched_reaction_prologue(reaction, worker);
-    #endif
-
-    LF_PRINT_DEBUG("Worker %d Execute Reaction", worker);
-    tracepoint_reaction_starts(reaction, worker);
-    ((self_base_t*) reaction->self)->current_tag = current_tag;
-    ((self_base_t*) reaction->self)->executing_reaction = reaction;
-    reaction->function(reaction->self);
-    ((self_base_t*) reaction->self)->executing_reaction = NULL;
-    tracepoint_reaction_ends(reaction, worker); 
-    LF_PRINT_DEBUG("Worker %d Finished Reaction", worker);
+    #endif        
+
+    bool violation = false;
+    if (reaction->deadline >= 0LL) {
+        // Get the current physical time.
+        instant_t physical_time = lf_time_physical();
+        // Check for deadline violation.
+        if (reaction->deadline == 0 || physical_time > current_tag.time + reaction->deadline) {
+            // Deadline violation has occurred.
+            violation = true;
+            // Invoke the local handler, if there is one.
+            reaction_function_t handler = reaction->deadline_violation_handler;
+            if (handler != NULL) {
+                (*handler)(reaction->self);
+            }
+        }
+    }
+    
```

Comment:
@edwardalee This is rather big change that was necessary. We have to check for deadline violation in the very end. Because we might stall in `_lf_sched_reaction_prologue` waiting for a mutex.